examples/scan_iterator.py

```diff
@@ -0,0 +1,103 @@
+"""
+Scan/Discovery Async Iterator
+--------------
+
+Example showing how to scan for BLE devices using async iterator instead of callback function
+
+Created on 2023-07-07 by bojanpotocnik <info@bojanpotocnik.com>
+
+"""
+import asyncio
+import contextlib
+from typing import AsyncIterator, Tuple
+
+from bleak import BleakScanner, BLEDevice, AdvertisementData
+
+
+async def normal_usage():
+    n = 5
+    print(f"Scanning for {n} devices...")
+    async with BleakScanner() as scanner:
+        async for bd, ad in scanner.observe():
+            print(f"{n}. {bd!r} with {ad!r}")
+            n -= 1
+            if n == 0:
+                break
+
+    n = 6
+    print(f"\nScanning for a device with name longer than {n} characters...")
+    async with BleakScanner() as scanner:
+        async for bd, ad in scanner.observe():
+            found = len(bd.name or "") > n or len(ad.local_name or "") > n
+            print(f"Found{' it' if found else ''} {bd!r} with {ad!r}")
+            if found:
+                break
+
+
+async def deterministic_cleanup():
```

Comment:
I would like to merge https://github.com/hbldh/bleak/pull/1367 first, then I think we can avoid the need for deterministic cleanup.

Reply:
I will wait for https://github.com/hbldh/bleak/pull/1367 to merge - then we also avoid dilemmas like https://github.com/hbldh/bleak/pull/1361#discussion_r1267361215

Reply:
Removed that code from the example. I could not read all the discussions here in the past half year, but if there are users hitting this problem in some other use cases, this info about proper cleanup can be added to some other (or separate) example.